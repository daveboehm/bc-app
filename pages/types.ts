// --------------
// Private types
// --------------

type ImageType = {
    height: number;
    img?: HTMLElement;
    title: string;
    url: string;
    width: number;
}

type UrlType = {
    value: string;
    // @TODO: confirm available options for link target attr
    target: "_blank" | "parent" | "self";
}

// --------------
// External types
// --------------

export type EntryType = {
    description?: string;
    id: number;
    image: ImageType;
    title: string;
    url: UrlType;
}

export type EntriesPropsType = {
    entryIds: (string | undefined)[];
    hydratedEntries: (EntryType | undefined)[];
}

export type PaginationProps = {
    currentPage: number;
    itemsPerPage?: number;
    handlePageChange: (_pageNumber: number) => void;
    totalItems: number;
}