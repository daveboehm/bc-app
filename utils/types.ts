// --------------
// Private types
// --------------

type ImageType = {
    height: number;
    img?: HTMLElement;
    title: string;
    url: string;
    width: number;
}

type UrlType = {
    value: string;
    target: "_blank" | "_parent" | "_self" | "_top";
}

// --------------
// External types
// --------------

export type EntryType = {
    description?: string;
    id: number;
    image: ImageType;
    title: string;
    url: UrlType;
}

export type EntriesPropsType = {
    entryIds: (string | undefined)[];
    hydratedEntries: (EntryType | undefined)[];
}

export type PaginationProps = {
    currentPage: number;
    itemsPerPage?: number;
    handlePageChange: (_pageNumber: number) => void;
    totalItems: number;
}