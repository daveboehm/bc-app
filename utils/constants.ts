export const ENTRIES_URL = 'https://www.bigcommerce.com/actions/bcCore/interview/getShowcaseEntryIds'

export const ENTRY_URL = 'https://www.bigcommerce.com/actions/bcCore/interview/getShowcaseEntryById'

export const ITEMS_PER_PAGE = 10