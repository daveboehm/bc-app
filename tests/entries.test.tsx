import React from 'react'
import { cleanup, render, screen, RenderResult } from "@testing-library/react";
import { EntriesPropsType } from '../utils/types';
import { MOCK_ENTRIES_PAGE_PROPS } from './__fixtures__/mock-data';
import Entries from '../pages/entries';
import { ITEMS_PER_PAGE } from '../utils/constants';

describe("<EntriesPage />", () => {

    describe("happy path", () => {
        let props: EntriesPropsType
        let view: RenderResult

        beforeEach(() => {
            props = MOCK_ENTRIES_PAGE_PROPS
            view = render(<Entries {...props} />)
        })

        afterEach(cleanup)

        it("should match snapshot", () => {
            expect(view.container).toBeVisible()
            expect(view.container).toMatchSnapshot()
        })

        it(`should display ${ITEMS_PER_PAGE}`, () => {
            const entriesOnPage = screen.getAllByTestId(/entry-item/)
            expect(entriesOnPage).toHaveLength(ITEMS_PER_PAGE)
        })
    })
})