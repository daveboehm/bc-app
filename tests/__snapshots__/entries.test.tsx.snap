// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EntriesPage /> happy path should match snapshot 1`] = `
<div>
  <main
    class="main"
  >
    <span
      class="backLink"
    >
      Back to Home
    </span>
    <hr
      class="hr"
    />
    <h1
      class="pageHeader"
    >
      All Entries (
      25
      )
    </h1>
    <ul
      class="entries"
    >
      <li
        class="entryItem"
        data-testid="entry-item"
      >
        <a
          class="entryItemLink"
          href="https://www.heatpresszone.com/"
          target="_blank"
          title="https://www-cdn.bigcommerce.com/assets/Heat-Press-Zone.png"
        >
          <div
            class="entryImage"
          >
            <img
              alt="Heat Press Zone"
              src="https://www-cdn.bigcommerce.com/assets/Heat-Press-Zone.png"
            />
          </div>
          <h2
            class="entryTitle"
          >
            Heat Press Zone
          </h2>
        </a>
      </li>
      <li
        class="entryItem"
        data-testid="entry-item"
      >
        <a
          class="entryItemLink"
          href="https://www.heatpresszone.com/"
          target="_blank"
          title="https://www-cdn.bigcommerce.com/assets/Heat-Press-Zone.png"
        >
          <div
            class="entryImage"
          >
            <img
              alt="Heat Press Zone"
              src="https://www-cdn.bigcommerce.com/assets/Heat-Press-Zone.png"
            />
          </div>
          <h2
            class="entryTitle"
          >
            Heat Press Zone
          </h2>
        </a>
      </li>
      <li
        class="entryItem"
        data-testid="entry-item"
      >
        <a
          class="entryItemLink"
          href="https://www.heatpresszone.com/"
          target="_blank"
          title="https://www-cdn.bigcommerce.com/assets/Heat-Press-Zone.png"
        >
          <div
            class="entryImage"
          >
            <img
              alt="Heat Press Zone"
              src="https://www-cdn.bigcommerce.com/assets/Heat-Press-Zone.png"
            />
          </div>
          <h2
            class="entryTitle"
          >
            Heat Press Zone
          </h2>
        </a>
      </li>
      <li
        class="entryItem"
        data-testid="entry-item"
      >
        <a
          class="entryItemLink"
          href="https://www.heatpresszone.com/"
          target="_blank"
          title="https://www-cdn.bigcommerce.com/assets/Heat-Press-Zone.png"
        >
          <div
            class="entryImage"
          >
            <img
              alt="Heat Press Zone"
              src="https://www-cdn.bigcommerce.com/assets/Heat-Press-Zone.png"
            />
          </div>
          <h2
            class="entryTitle"
          >
            Heat Press Zone
          </h2>
        </a>
      </li>
      <li
        class="entryItem"
        data-testid="entry-item"
      >
        <a
          class="entryItemLink"
          href="https://www.heatpresszone.com/"
          target="_blank"
          title="https://www-cdn.bigcommerce.com/assets/Heat-Press-Zone.png"
        >
          <div
            class="entryImage"
          >
            <img
              alt="Heat Press Zone"
              src="https://www-cdn.bigcommerce.com/assets/Heat-Press-Zone.png"
            />
          </div>
          <h2
            class="entryTitle"
          >
            Heat Press Zone
          </h2>
        </a>
      </li>
      <li
        class="entryItem"
        data-testid="entry-item"
      >
        <a
          class="entryItemLink"
          href="https://www.heatpresszone.com/"
          target="_blank"
          title="https://www-cdn.bigcommerce.com/assets/Heat-Press-Zone.png"
        >
          <div
            class="entryImage"
          >
            <img
              alt="Heat Press Zone"
              src="https://www-cdn.bigcommerce.com/assets/Heat-Press-Zone.png"
            />
          </div>
          <h2
            class="entryTitle"
          >
            Heat Press Zone
          </h2>
        </a>
      </li>
      <li
        class="entryItem"
        data-testid="entry-item"
      >
        <a
          class="entryItemLink"
          href="https://www.heatpresszone.com/"
          target="_blank"
          title="https://www-cdn.bigcommerce.com/assets/Heat-Press-Zone.png"
        >
          <div
            class="entryImage"
          >
            <img
              alt="Heat Press Zone"
              src="https://www-cdn.bigcommerce.com/assets/Heat-Press-Zone.png"
            />
          </div>
          <h2
            class="entryTitle"
          >
            Heat Press Zone
          </h2>
        </a>
      </li>
      <li
        class="entryItem"
        data-testid="entry-item"
      >
        <a
          class="entryItemLink"
          href="https://www.heatpresszone.com/"
          target="_blank"
          title="https://www-cdn.bigcommerce.com/assets/Heat-Press-Zone.png"
        >
          <div
            class="entryImage"
          >
            <img
              alt="Heat Press Zone"
              src="https://www-cdn.bigcommerce.com/assets/Heat-Press-Zone.png"
            />
          </div>
          <h2
            class="entryTitle"
          >
            Heat Press Zone
          </h2>
        </a>
      </li>
      <li
        class="entryItem"
        data-testid="entry-item"
      >
        <a
          class="entryItemLink"
          href="https://www.heatpresszone.com/"
          target="_blank"
          title="https://www-cdn.bigcommerce.com/assets/Heat-Press-Zone.png"
        >
          <div
            class="entryImage"
          >
            <img
              alt="Heat Press Zone"
              src="https://www-cdn.bigcommerce.com/assets/Heat-Press-Zone.png"
            />
          </div>
          <h2
            class="entryTitle"
          >
            Heat Press Zone
          </h2>
        </a>
      </li>
      <li
        class="entryItem"
        data-testid="entry-item"
      >
        <a
          class="entryItemLink"
          href="https://www.heatpresszone.com/"
          target="_blank"
          title="https://www-cdn.bigcommerce.com/assets/Heat-Press-Zone.png"
        >
          <div
            class="entryImage"
          >
            <img
              alt="Heat Press Zone"
              src="https://www-cdn.bigcommerce.com/assets/Heat-Press-Zone.png"
            />
          </div>
          <h2
            class="entryTitle"
          >
            Heat Press Zone
          </h2>
        </a>
      </li>
    </ul>
    <hr
      class="hr"
    />
    <div
      class="paginationContainer"
    >
      <div
        class="buttonsContainer"
      >
        <button
          class="pageButtonInactive"
          data-testid="pagination-prev-button"
          disabled=""
          type="button"
        >
          Prev
        </button>
        <button
          class="pageButtonActive"
          data-testid="pagination-button"
          type="button"
        >
          1
        </button>
        <button
          class="pageButtonInactive"
          data-testid="pagination-button"
          type="button"
        >
          2
        </button>
        <button
          class="pageButtonInactive"
          data-testid="pagination-button"
          type="button"
        >
          3
        </button>
        <button
          class="pageButtonInactive"
          data-testid="pagination-next-button"
          type="button"
        >
          Next
        </button>
      </div>
    </div>
  </main>
</div>
`;
